### PR TITLE
Issue #4533 Hard close from Dispatcher

### PR DESCRIPTION
**Pull request by gregw**
#4533 Do hard close from Dispatcher so response wrappers may intercept close.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

File: jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DispatcherTest.java [hunk 1]
```diff
@@ -373,6 +377,98 @@ public void testForwardFilterToRogerServlet() throws Exception
...
+                @Override
+                public void write(byte[] b, int off, int len) throws IOException
+                {
+                    buffer.write(b, off, len);
+                }
+
+                @Override
+                public void close() throws IOException
+                {
+                    buffer.close();
```
Comment (sbordet) [inline jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DispatcherTest.java:458]: I have commented out this line `buffer.close()`, and the test still passes.

I'm not sure at this point what are we really testing? I would have expected the test to fail if `buffer.close()` was commented out?

If the point of the test is to verify that the wrapped `close()` is invoked, can you make it explicit with a `boolean` or a `CountDownLatch`?

File: jetty-server/src/main/java/org/eclipse/jetty/server/Dispatcher.java [hunk 2]
```diff
@@ -219,8 +219,18 @@ protected void forward(ServletRequest request, ServletResponse response, Dispatc
...
+                {
+                    try
+                    {
+                        response.getOutputStream().close();
+                    }
+                    catch (IllegalStateException e)
+                    {
+                        response.getWriter().close();
+                    }
+                }
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/Dispatcher.java:233]: There remains only one other usage of `softClose()`. Is it really necessary, or `softClose()` can be removed?

File: jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DispatcherTest.java (hunk 1 above)
Comment (gregw) [reply jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DispatcherTest.java:458]: If you remove the method or call super.close() then the test fails.   The test correctly reproduces the failure produced by the client when the fix is not in place.

File: jetty-server/src/main/java/org/eclipse/jetty/server/Dispatcher.java (hunk 2 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/Dispatcher.java:233]: The concept of a softClose is still needed by sendError - as we need to be able to treat the output as closed after the sendError for the remaining portion of a dispatch, only to reopen to send the actual response in the errorHandler or ERROR dispatch.    It was wrong to apply that concept to dispatch.forward, as there is no reopen semantic needed after a forward, plus there is a wrapping semantic that we broke.

**Review by sbordet — APPROVED** [review]

**Pull request merged**